jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/content/ByteChannelContentSource.java

```diff
@@ -0,0 +1,303 @@
+//
+// ========================================================================
+// Copyright (c) 1995 Mort Bay Consulting Pty Ltd and others.
+//
+// This program and the accompanying materials are made available under the
+// terms of the Eclipse Public License v. 2.0 which is available at
+// https://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
+// which is available at https://www.apache.org/licenses/LICENSE-2.0.
+//
+// SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+// ========================================================================
+//
+
+package org.eclipse.jetty.io.content;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.nio.channels.ByteChannel;
+import java.nio.channels.ClosedChannelException;
+import java.nio.channels.SeekableByteChannel;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.StandardOpenOption;
+import java.util.Objects;
+
+import org.eclipse.jetty.io.ByteBufferPool;
+import org.eclipse.jetty.io.Content;
+import org.eclipse.jetty.io.RetainableByteBuffer;
+import org.eclipse.jetty.util.BufferUtil;
+import org.eclipse.jetty.util.ExceptionUtil;
+import org.eclipse.jetty.util.IO;
+import org.eclipse.jetty.util.thread.AutoLock;
+import org.eclipse.jetty.util.thread.SerializedInvoker;
+
+/**
+ * <p>A {@link Content.Source} backed by a  {@link ByteChannel}.
+ * Any calls to {@link #demand(Runnable)} are immediately satisfied.</p>
+ */
+public class ByteChannelContentSource implements Content.Source
+{
+    private final AutoLock lock = new AutoLock();
+    private final SerializedInvoker _invoker = new SerializedInvoker();
+    private final ByteBufferPool.Sized _byteBufferPool;
+    private ByteChannel _byteChannel;
+    private final long _offset;
+    private final long _length;
+    private RetainableByteBuffer _buffer;
+    private long _totalRead;
+    private Runnable demandCallback;
+    private Content.Chunk _terminal;
+
+    public ByteChannelContentSource(SeekableByteChannel seekableByteChannel, long offset, long length)
+    {
+        this(new ByteBufferPool.Sized(null), seekableByteChannel, offset, length);
+    }
+
+    public ByteChannelContentSource(ByteBufferPool.Sized byteBufferPool, SeekableByteChannel seekableByteChannel, long offset, long length)
+    {
+        this(byteBufferPool, (ByteChannel)seekableByteChannel, offset, length);
+        if (offset > 0 && seekableByteChannel != null)
```

Comment:
We do not know if we receive a `seekableByteChannel` with `position=0`, so I think we should enforce it anyway:

```suggestion
        if (offset >= 0 && seekableByteChannel != null)
```